ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -31,6 +45,255 @@ let tdb_tool = !Xapi_globs.tdb_tool
+      ; password_expired= Int32.logand user_account_control 8388608l <> 0l
+      }
+
+  let query_user sid =
+    let query = Printf.sprintf "(&(objectClass=user)(objectSid=%s))" sid in
+    let* stdout =
+      call_ldap query <!> fun _ -> generic_ex "ldap query failed: '%s'" query
+    in
+    parse_user stdout <!> generic_ex "%s"
+end
+
+module Wbinfo = struct
+  let can_resolve_krbtgt () =
+    match call_wbinfo ["-n"; krbtgt] with Ok _ -> true | Error _ -> false
+
+  let sid_of_name name =
+    (* example:
+     *
+     * $ wbinfo -n user@domain.net
+       S-1-2-34-... SID_USER (1)
+     * $ wbinfo -n DOMAIN\user
+       # similar output *)
+    let args = ["-n"; name] in
+    let* stdout =
+      call_wbinfo args <!> fun _ ->
+      generic_ex "'wbinfo %s' failed" (String.concat " " args)
+    in
+    match String.split_on_char ' ' stdout with
+    | sid :: _ ->
+        Ok (String.trim sid)
+    | [] ->
+        Error
+          (generic_ex "unable to find SID in output of 'wbinfo %s'"
+             (String.concat " " args))
+
+  let name_of_sid =
+    (* example:
+     * $ wbinfo -s S-1-5-21-3143668282-2591278241-912959342-502
+       CONNAPP\krbtgt 1 *)
+    (* we need to drop the last digit *)
+    let regex = Re.Perl.(compile (re {|^([^\s].*)\ \d+\s*$|})) in
+    let get_regex_match x =
+      Option.bind (Re.exec_opt regex x) (fun g ->
+          try Some (Re.Group.get g 1) with _ -> None)
+    in
+    fun sid ->
+      let args = ["-s"; sid] in
+      let* stdout =
+        call_wbinfo args <!> fun _ ->
+        generic_ex "'wbinfo %s' failed" (String.concat " " args)
+      in
+      match get_regex_match stdout with
+      | None ->
+          Error
+            (generic_ex "parsing 'wbinfo %s' failed" (String.concat " " args))
+      | Some x ->
+          Ok x
+
+  let gid_of_sid sid =
+    let args = ["-Y"; sid] in
+    let* stdout =
+      call_wbinfo args <!> fun _ ->
+      generic_ex "'wbinfo %s' failed" (String.concat " " args)
+    in
+    try Ok (String.trim stdout |> int_of_string)
+    with _ ->
+      Error (generic_ex "parsing 'wbinfo %s' failed" (String.concat " " args))
+
+  let user_domgroups sid =
+    (* example:
+     *
+     * $ wbinfo --user-domgroups S-1-2-34-...
+       S-1-2-34-...
+       S-1-5-21-...
+       ... *)
+    let args = ["--user-domgroups"; sid] in
+    let* stdout =
+      call_wbinfo args <!> fun _ ->
+      generic_ex "'wbinfo %s' failed" (String.concat " " args)
+    in
+    Ok (String.split_on_char '\n' stdout |> List.map String.trim)
+
+  let uid_of_sid sid =
+    let args = ["-S"; sid] in
+    let* stdout =
+      call_wbinfo args <!> fun _ ->
+      generic_ex "'wbinfo %s' failed" (String.concat " " args)
+    in
+    try Ok (String.trim stdout |> int_of_string)
+    with _ ->
+      Error (generic_ex "parsing 'wbinfo %s' failed" (String.concat " " args))
+
+  type uid_info = {user_name: string; uid: int; gid: int; gecos: string}
```

Comment:
It is really good to extract information into struct (types) like `user` and `uid_info` 👍 , 
Just a question, should the type be defined inside the modules, like `uid_info`, 
as `uid_info` does not looks only specific to the `wbinfo` command.

Reply:
It is specific to wbinfo in the sense that we obtain the information from wbinfo (so I believe the `Wbinfo` module is in the right place for it)